docs/design/generic-enhanced-purls.md

```diff
@@ -0,0 +1,233 @@
+# Support for different purl types in the generic artifact fetcher
+
+The generic artifact package manager is being added to Cachi2 as a means for users to introduce files that do not belong to traditional package manager ecosystems (e.g. pip, npm, golang) to their hermetic container builds. Since Cachi2 does not have any extra information about the file that's being fetched, the purls are always reported as [pkg:generic](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#generic).
+
+There are use cases that would benefit from more accurate purls, though, such as the recent Maven artifacts [proposal](https://github.com/containerbuildsystem/cachi2/pull/663). Considering that the purl specification already identifies several types of packages that don't fit into traditional package manager (e.g. github, docker, huggingface; see the [purl types spec](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst) for more info), this proposal builds on top of the fundamentals of the generic fetcher to provide an extensible mechanism that would allow Cachi2 to fetch files from specific sources and report them with matching purl types.
+
+## Consuming input purls
+
+A purl is simply a way to represent a unique package and it's location on the internet. By allowing users to specify a purl as part of a package definition in the lockfile, Cachi2 can verify that it actually resolves to a valid download location, and that it adheres to the fundamentals of the purl spec. Note that there's a great variety of purl types, and resolving them will involve type-specific logic, which means we'll only be able to support a small subset of purl types.
+
+This is a summary of how the resolution of a single package will look like:
+
+- Instead of consuming a `download_url`, the package will specify a `purl` attribute in the lockfile
+- Parse the purl by using the [packageurl-python](https://github.com/package-url/packageurl-python) library
+- Validate that the purl is within the supported types
```

Comment:
Earlier you mentioned that purls are reported as generic. Would Cachi2 also support a generic purl as input? If the supported types mentioned below are not used, would Cachi2 try to just fall back to fetch a generic artifact?

Reply:
I think we should keep the current lockfile option of provinding `donwload_url` and `checksums` for the `pkg:generic` cases, but we can also support consuming purls as input for the sake of consistency.

As for the types, I think we need to be very explicit about which ones are supported, and anything that falls outside of that means a failed request.

Reply:
So the only way that users would be able to prefetch in a way that produces a generic purl would be through the current lockfile option? There is no proposal to add a purl-supported generic package fetching?

Reply:
There isn't, but we can add one. Do you think it's worth doing it now?

Reply:
No, I did not write anything about consuming input `pkg:generic` purls here because this is already supported by the current lockfile. So, initally, Cachi2 would not consume `pkg:generic` purls and for this reason, users wouldn't be able to produce `pkg:generic` SBOM components by providing input purls.

The mechanism to extend support to any purl type it is already described here, though, so supporting input `pkg:generic` is trivial (and we probably should, for the sake of consistency).